cmake/DefaultTargetProperties.cmake

```diff
@@ -0,0 +1,129 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2013-2022 The Foundry Visionmongers Ltd
+
+function(set_default_target_properties target_name)
+    #-------------------------------------------------------------------
+    # C++ standard
+
+    # Minimum C++ standard
+    target_compile_features(${target_name} PRIVATE cxx_std_14)
+
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        # Ensure the proposed compiler supports our minimum C++
+        # standard.
+        CXX_STANDARD_REQUIRED ON
+        # Disable compiler extensions. E.g. use -std=c++11  instead of
+        # -std=gnu++11.  Helps limit cross-platform issues.
+        CXX_EXTENSIONS OFF
+    )
+
+    #-------------------------------------------------------------------
+    # Compiler warnings
+
+    include(CompilerWarnings)
+    set_default_compiler_warnings(${target_name})
+
+
+    #-------------------------------------------------------------------
+    # Potentially enable sanitizers.
+    include(Sanitizers)
+    enable_sanitizers(project_options)
+
+
+    #-------------------------------------------------------------------
+    # Symbol visibility
+
+    # Hide symbols from this library by default.
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        C_VISIBILITY_PRESET hidden
+        CXX_VISIBILITY_PRESET hidden
+        VISIBILITY_INLINES_HIDDEN YES
+    )
+
+    # Hide all symbols from external statically linked libraries.
+    if (CMAKE_CXX_COMPILER_ID MATCHES ".*Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+        target_link_options(${target_name} PRIVATE -Wl,--exclude-libs,ALL)
+    endif ()
+
+    # Use C++11 ABI by default, but allow disabling.
+    if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND
+        CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 5.0)
+        option(GLIBCXX_USE_CXX11_ABI "For gcc, use the new C++11 library ABI" ON)
+        if (GLIBCXX_USE_CXX11_ABI)
+            target_compile_definitions(${target_name} PRIVATE _GLIBCXX_USE_CXX11_ABI=1)
+        else ()
+            target_compile_definitions(${target_name} PRIVATE _GLIBCXX_USE_CXX11_ABI=0)
+        endif ()
+    endif ()
+
+    # TODO(DF): POSITION_INDEPENDENT_CODE required for Python module? See OCIO.
+
+    #-------------------------------------------------------------------
+    # RPATH wrangling.
+
+    # With the 'usual' install path structure in mind, search from the
+    # bin directory (i.e. a binary loading a dynamic library) and then
+    # from the current directory (i.e. dynamic library loading another
+    # dynamic library).
+    if (APPLE)
+        set(rpath "@loader_path/../${CMAKE_INSTALL_LIBDIR};@loader_path")
+    else ()
+        set(rpath "$ORIGIN/../${CMAKE_INSTALL_LIBDIR};$ORIGIN")
+    endif ()
+
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
```

Comment:
Majority of these I wouldn't even specify unless they're absolutely needed.

Reply:
Hmm I wanted to be explicit since I don't know off the top of my head what the defaults are. Is there any danger in specifying them?

In fact I had considered we may set `SKIP_BUILD_RPATH TRUE` if we ensure all tests run from an install rather than the build directory, but that may be annoying for TDD development.

Reply:
How slow is the `install` step that you'd need? Would it practically add that much to a TDD loop?

Reply:
Probably negligible to be fair, it'll only ever be a relatively tiny project. It'll be a little bit more dev work to set up the test binaries to link to an installed library rather than the build target. But we should really test our installed lib can be linked and used, anyway.

Reply:
IMO, it's useful to be able to run unittests from a build folder, rather than an install folder. (In terms of saving developer time.)
However, it can also be said that it's useful to run unittests on the install folder, as you're testing what the consumer gets.

Reply:
Tentatively removed build tree RPATH and added a TODO re. testing strategy in 994b3521